From e306e53364ce3e21554a6a1e907bc433ba52b4cd Mon Sep 17 00:00:00 2001
From: Anurag Kumar Vulisha <anurag.kumar.vulisha@xilinx.com>
Date: Thu, 23 Jan 2020 19:50:50 +0530
Subject: [PATCH 0999/1851] usb: xhci-plat: Add remote wakeup support

commit a87dd12d2b162158c2e034ed10ca64f781b8ed8b from
https://github.com/Xilinx/linux-xlnx.git

This patch adds support for enabling remote wakeup capability
to the host controller

Signed-off-by: Anurag Kumar Vulisha <anuragku@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Anurag Kumar Vulisha <anurag.kumar.vulisha@xilinx.com>
Signed-off-by: Piyush Mehta <piyush.mehta@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/usb/host/xhci-plat.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/host/xhci-plat.c b/drivers/usb/host/xhci-plat.c
index 3bfac0b834dd..e9873ce3aec8 100644
--- a/drivers/usb/host/xhci-plat.c
+++ b/drivers/usb/host/xhci-plat.c
@@ -243,6 +243,9 @@ static int xhci_plat_probe(struct platform_device *pdev)
 			return ret;
 	}
 
+	/* Set the controller as wakeup capable */
+	device_set_wakeup_capable(&pdev->dev, true);
+
 	pm_runtime_set_active(&pdev->dev);
 	pm_runtime_enable(&pdev->dev);
 	pm_runtime_get_noresume(&pdev->dev);
@@ -440,10 +443,12 @@ static int __maybe_unused xhci_plat_suspend(struct device *dev)
 
 #if IS_ENABLED(CONFIG_USB_DWC3_OF_SIMPLE)
 	/* Inform dwc3 driver about the device wakeup capability */
-	if (device_may_wakeup(dev))
+	if (device_may_wakeup(&hcd->self.root_hub->dev)) {
+		enable_irq_wake(hcd->irq);
 		dwc3_host_wakeup_capable(dev, true);
-	else
+	} else {
 		dwc3_host_wakeup_capable(dev, false);
+	}
 #endif
 
 	/*
@@ -483,6 +488,9 @@ static int __maybe_unused xhci_plat_runtime_resume(struct device *dev)
 	struct usb_hcd  *hcd = dev_get_drvdata(dev);
 	struct xhci_hcd *xhci = hcd_to_xhci(hcd);
 
+	if (device_may_wakeup(&hcd->self.root_hub->dev))
+		disable_irq_wake(hcd->irq);
+
 	return xhci_resume(xhci, 0);
 }
 
-- 
2.31.1

